File: 0174-dungeon-game/0174-dungeon-game.cpp
class Solution {
public:
    //int calculateMinimumHP(vector<vector<int>>& dungeon) {
        //find max no. which will be -ve
        //hame max -ve no. nikalna hai
        //mtlb sabse chota dis
        //-ve no. found then return 1+(+ve of(-no.))
        
    //     int n=dungeon.size();
    //     int m=dungeon[0].size();
    //     int ans=INT_MAX;
    //     vector<vector<int>>dist(n,vector<int>(m,INT_MIN));//we have to store max health value
    //     dist[0][0]=dungeon[0][0];
    //     queue<pair<int,pair<int,int>>>q;///d r c
    //     q.push({dungeon[0][0],{0,0}});
    //     vector<pair<int,int>>dir={{0,1},{1,0}};
    //     int ans1=INT_MAX;
    //     int ans2=INT_MAX;
    //     while(!q.empty()){
    //         int d=q.front().first;
    //         int r=q.front().second.first;
    //         int c=q.front().second.second;
    //         q.pop();
    //         if(r==n-1 && c==m-1){
    //             ans=min(ans,d);
    //         }
    //         if(r==n-2 && c==m-1){
    //             ans1=min(ans1,d);
    //         }
    //         if(r==n-1 && c==m-2){
    //             ans2=min(ans2,d);
    //         }
    //         for(int i=0;i<2;i++){
    //             int newr=r+dir[i].first;
    //             int newc=c+dir[i].second;
    //             if(newr>=0 && newr<n && newc>=0 && newc<m && dist[newr][newc]<d+dungeon[newr][newc]){
    //                     dist[newr][newc]=d+dungeon[newr][newc];
    //                     q.push({dist[newr][newc],{newr,newc}});
    //             }
    //         }
    //     }
    //     if(ans>=0)return 1;
    //     else if(ans1<0 && ans1<ans2 && dist[n-1][m-1]>0) return -1*ans1+1;
    //     else if(ans2<0 && ans2<ans1 && dist[n-1][m-1]>0) return -1*ans2+1;
    //     return 1+ (-1*ans);
        
    // }
    //wrong ans
    /*
    Incorrect Dynamic Programming Approach: The problem is not about maximizing health but about ensuring the minimum amount of health required to survive at any given cell. Your approach is trying to maximize the health gained, which leads to incorrect calculations.

Priority Queue Not Needed: You are trying to solve this with a BFS-like approach, but the problem requires dynamic programming to track the minimum health needed to reach the bottom-right cell starting from the bottom-right and moving backward.

Misinterpretation of Conditions: The conditions like ans1 and ans2 are not relevant to this problem. The problem requires us to ensure that we always have positive health at every cell along the way from start to end.
    */

    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int n = dungeon.size();
        int m = dungeon[0].size();
        
        // DP table to store minimum health needed at each cell
        vector<vector<int>> dp(n, vector<int>(m, INT_MAX));
        
        // Base case: The knight must have at least 1 health point to enter the princess's room
        dp[n-1][m-1] = max(1, 1 - dungeon[n-1][m-1]);

        // Fill the last row (from bottom to top)
        for (int i = n - 2; i >= 0; --i) {
            dp[i][m-1] = max(1, dp[i+1][m-1] - dungeon[i][m-1]);
        }

        // Fill the last column (from right to left)
        for (int j = m - 2; j >= 0; --j) {
            dp[n-1][j] = max(1, dp[n-1][j+1] - dungeon[n-1][j]);
        }

        // Fill the rest of the table
        for (int i = n - 2; i >= 0; --i) {
            for (int j = m - 2; j >= 0; --j) {
                int down = max(1, dp[i+1][j] - dungeon[i][j]);
                int right = max(1, dp[i][j+1] - dungeon[i][j]);
                dp[i][j] = min(down, right);
            }
        }

        // The result is the minimum health needed at the starting point
        return dp[0][0];
    }

};

File: 0174-dungeon-game/0174-dungeon-game.java
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        // Check if the dungeon grid is null or empty, in which case return 1 (minimum HP to start).
        if(dungeon == null || dungeon.length == 0 || dungeon[0].length == 0) return 1;
        
        int N = dungeon.length; // Number of rows in the dungeon.
        int M = dungeon[0].length; // Number of columns in the dungeon.
        
        // Initialize the DP table (2D array) to store the minimum HP required at each position.
        int[][] dp = new int[N][M];
        
        // Base case: the minimum HP needed at the princess's room (bottom-right corner) should be 1.
        // If dungeon[N-1][M-1] is positive, we subtract it from 1; if it's negative, we adjust HP accordingly.
        dp[N - 1][M - 1] = 1 - dungeon[N - 1][M - 1];
        dp[N - 1][M - 1] = dp[N - 1][M - 1] <= 0 ? 1 : dp[N - 1][M - 1];  // Ensure at least 1 HP is needed.

        // Traverse the dungeon grid from bottom-right to top-left.
        for(int i = N - 1; i >= 0; --i){ // Loop through rows.
            for(int j = M - 1; j >= 0; --j){ // Loop through columns.
                if(i == N - 1 && j == M - 1) continue; // Skip the princess's room, already initialized.

                // Calculate the minimum HP required to reach the bottom or right cell.
                // HP_D: HP needed if moving down, or Integer.MAX_VALUE if out of bounds.
                // HP_R: HP needed if moving right, or Integer.MAX_VALUE if out of bounds.
                int HP_D = (i + 1 == N) ? Integer.MAX_VALUE : dp[i + 1][j] - dungeon[i][j];
                int HP_R = (j + 1 == M) ? Integer.MAX_VALUE : dp[i][j + 1] - dungeon[i][j];
                
                // The minimum HP required at position (i, j) is the minimum between HP_D and HP_R.
                int HP = Math.min(HP_D, HP_R);
                
                // Ensure that HP is at least 1, meaning the player will have at least 1 HP to move forward.
                dp[i][j] = HP <= 0 ? 1 : HP;
            }    
        }
        
        // Return the minimum HP required to reach the princess starting from the top-left corner.
        return dp[0][0];
    }
}
